python312Packages.pygerber: init at 2.4.2, pytest-lsp: init at 0.4.3

clemjvdm opened the pull request:
<!--
Added the pygerber package, a python library that provides tools for working with gerber files.
[pygerber](https://github.com/Argmaster/pygerber)
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

steeleduncan commented on pkgs/development/python-modules/pygerber/default.nix:
```diff
@@ -0,0 +1,91 @@
+
+  pythonImportsCheck = [ "pygerber" ];
+
+  meta = with lib; {
```
`with lib` is deprecated #208242

steeleduncan commented on pkgs/development/python-modules/pygerber/default.nix:
```diff
@@ -0,0 +1,91 @@
+}:
+
+let
+  pillow_10_4 = pillow.overrideAttrs (prev: {
```
does it make sense to supply some variables for the `pillow` dependency as well, or is it likely to be at 10.4 forever?

Its personal of course, but I'd suggest not having the version in the variable name

steeleduncan commented on maintainers/maintainer-list.nix:
```diff
@@ -4165,6 +4165,12 @@
     githubId = 71959829;
     name = "Cleeyv";
   };
+  clemjvdm = {
```
please split the maintainer list change into a separate commit, preceeding the commit that introduces the package

clemjvdm commented:
do I need to do anything else? Sorry it's my first PR so I'm not sure. @steeleduncan

steeleduncan reviewed (commented):
I am not familiar enough with python derivations to leave an approving review, but I certainly wouldn't want to block it

If you can find someone who has merged some python commits recently and ping them for a review that is probably the best approach to get it merged

clemjvdm commented:
> I am not familiar enough with python derivations to leave an approving review, but I certainly wouldn't want to block it
> 
> If you can find someone who has merged some python commits recently and ping them for a review that is probably the best approach to get it merged

@natsukium Hello, hope it's ok for me to ping you here. I opened this PR about 3 weeks ago and I wasn't sure if I had to do anything else to get it merged.

FliegendeWurst commented on pkgs/development/python-modules/pygerber/default.nix:
```diff
@@ -0,0 +1,91 @@
+    owner = "Argmaster";
+    repo = pname;
+    rev = "refs/tags/v${version}";
+    hash = "sha256-+s89tHNTNP+0VMilfDczbOBG/e3DuUBu0113US69KKM=";
```
```suggestion
    hash = "sha256-N+9I59WiWXSXr7RrPzb7GFSqfjrd0q51AzalNFV4xEQ=";
```

FliegendeWurst commented on pkgs/development/python-modules/pygerber/default.nix:
```diff
@@ -0,0 +1,91 @@
+buildPythonPackage rec {
+  pname = "pygerber";
+  version = "2.4.2";
+  format = "pyproject";
```
```suggestion
  pyproject = true;
```

FliegendeWurst commented on pkgs/development/python-modules/pygerber/default.nix:
```diff
@@ -0,0 +1,91 @@
+
+}:
+
+let
+  pillow_10_4 = pillow.overrideAttrs (prev: {
+    version = "10.4.0";
+    src = fetchFromGitHub {
+      owner = "python-pillow";
+      repo = "pillow";
+      rev = "refs/tags/10.4.0";
+      hash = "sha256-Q3yvFyJf4gFaVY9X4PlgCE2rU1l9NEXkHBaNZfoHhvU=";
+    };
+  });
+
+in
```
```suggestion
```
Upstream supports newer Pillow versions https://github.com/Argmaster/pygerber/blob/9c2eeb4099822a9ef9a00938e43fef9d65dc638b/pyproject.toml#L58-L61

FliegendeWurst commented on pkgs/development/python-modules/pygerber/default.nix:
```diff
@@ -0,0 +1,91 @@
+    hash = "sha256-+s89tHNTNP+0VMilfDczbOBG/e3DuUBu0113US69KKM=";
+  };
+
+  nativeBuildInputs = [ poetry-core ];
```
```suggestion
  build-system = [ poetry-core ];
```

FliegendeWurst commented on pkgs/development/python-modules/pygerber/default.nix:
```diff
@@ -0,0 +1,91 @@
+  };
+
+  nativeBuildInputs = [ poetry-core ];
+  propagatedBuildInputs = [
```
```suggestion
  dependencies = [
```

FliegendeWurst commented on pkgs/development/python-modules/pygerber/default.nix:
```diff
@@ -0,0 +1,91 @@
+  pythonImportsCheck = [ "pygerber" ];
+
+  meta = {
+    description = "PyGerber is a Python implementation of Gerber X3/X2 format. It is based on Ucamco's The Gerber Layer Format Specification.";
```
```suggestion
    description = "Implementation of the Gerber X3/X2 format, based on Ucamco's The Gerber Layer Format Specification";
```

clemjvdm committed:
maintainers: add clemjvdm

clemjvdm commented:
@FliegendeWurst Just committed your suggestions, fixed a little error and squashed them, hopefully this is how I was supposed to do this. Anything else I should do?

FliegendeWurst commented:
Yes. So far you only activated the import check. It would be good to run the real test suite, if the project has one. Add pytestCheckHook (or another hook, depending on the framework) to nativeCheckInputs for that.

clemjvdm commented:
@FliegendeWurst I looked into it, but I can't test because `pytest-lsp` doesn't yet exist on nixpkgs :(

FliegendeWurst commented:
Here you go. Untested.

```
{
  lib,
  fetchPypi,
  buildPythonPackage,
  hatchling,
  pygls,
  pytestCheckHook,
  pytest-asyncio,
}:

buildPythonPackage rec {
  pname = "pytest-lsp";
  version = "0.4.3";
  pyproject = true;

  src = fetchPypi {
    inherit version;
    pname = "pytest_lsp";
    hash = "sha256-ND9r2i+qMg7V/Ld8lCDScDzlZdHRRP6CfjGYp9wpkRw=";
  };

  build-system = [
    hatchling
  ];

  dependencies = [
    pygls
    pytest-asyncio
  ];

  nativeCheckInputs = [
    pytestCheckHook
  ];

  pythonImportsCheck = [ "pytest_lsp" ];

  meta = {
    homepage = "https://github.com/swyddfa/lsp-devtools";
    changelog = "https://github.com/swyddfa/lsp-devtools/blob/develop/lib/pytest-lsp/CHANGES.md";
    description = "Pytest plugin for writing end-to-end tests for language servers";
    license = lib.licenses.mit;
    maintainers = with lib.maintainers; [ clemjvdm ];
  };
}
```

clemjvdm commented:
@FliegendeWurst Uhh sorry, I'm not sure how I should go about this?. Open a new PR with pytest_lsp, or just write this somewhere in this PR?

FliegendeWurst commented:
You can put it in this PR. Ideally it would be done in a commit before the pygerber package is added.

clemjvdm commented:
@FliegendeWurst Just to be sure before I do this, pygerber documentation states that to run tests one should do:
```
poetry shell
poe run-unit-tests
```
I assumed this wasn't the right way to do this here since this would imply letting poetry manage dependencies. I also tried just installing the required dependencies and doing `poe run-unit-tests` but this requires poetry and nix will only let me use `poetry-core`:
```
error: poetry was promoted to a top-level attribute, use poetry-core to build Python packages
```

Which is why I figured maybe using pytests (which poepoet refers to) directly was the right approach. But in hindsight I'm not sure.

clemjvdm commented:
I think either way we'll need `pytest-lsp`.

FliegendeWurst commented:
I think pytestCheckHook will work. You may need to add some more plugins though:

https://github.com/Argmaster/pygerber/blob/main/pyproject.toml#L84-L89

clemjvdm commented:
@FliegendeWurst Yeah I figured, I'll try that then. Thanks :)

clemjvdm committed:
python312Packages.pytest-lsp: init at 0.4.3

clemjvdm commented:
@FliegendeWurst New issue! Yay! I was able to get everything working with `pytest-lsp` (thanks again). But it seems pygerber's tests pull assets from github (which isn't allowed by the nix build phase).

Is there a good way of solving this? I presume maybe we can fetch them in nix and then patch the tests but I don't have time nor am I qualified to do that I think.

FliegendeWurst commented:
If it's only some tests you can disable them using disabledTests = [ .... ].

clemjvdm commented:
@FliegendeWurst Ok that worked! Do we care about skipped, xfailed and xpassed tests?
```
============= 427 passed, 2 skipped, 8 xfailed, 1 xpassed in 7.08s =============
```

Otherwise I'll commit.

FliegendeWurst commented:
That looks good to me :+1:

clemjvdm commented:
@FliegendeWurst All done. Thanks you sm for the help :)

FliegendeWurst commented on pkgs/development/python-modules/pygerber/default.nix:
```diff
@@ -0,0 +1,105 @@
+
+  src = fetchFromGitHub {
+    owner = "Argmaster";
+    repo = pname;
+    rev = "refs/tags/v${version}";
```
```suggestion
    repo = "pygerber;
    tag = "v${version}";
```
Avoid using pname #277994.
The tag parameter saves some boilerplate. (you may need to rebase if your branch does not support it)

FliegendeWurst commented on pkgs/development/python-modules/pygerber/default.nix:
```diff
@@ -0,0 +1,105 @@
+    filelock
+  ];
+
+  disabledTestPaths = [
+    "test/gerberx3/test_assets.py"
+    "test/gerberx3/test_language_server/tests.py"
+  ];
```
```suggestion
  disabledTestPaths = [
    # require network access
    "test/gerberx3/test_assets.py"
    "test/gerberx3/test_language_server/tests.py"
  ];
```
Always good to explain why certain tests are disabled.

clemjvdm committed:
python312Packages.pygerber: init at 2.4.2
